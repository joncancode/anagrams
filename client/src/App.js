import React, { Component } from 'react';
import './App.css';

//this is a class component that will be used for our frontend
//while i could have broken this into multiple components, i used a single one, since it all was based off of the single action of one button click
class App extends Component {
  //because we are going to do some configuring to the state, we need to use the constructor and super in this component
  constructor() {
    super();

    //the state will need to hold the input value as well as the total words in the dictionary text file.
    this.state = {
      inputVal: '',
      words: []
    };
  }

  //when the component mounts, it makes a fetch request to the backend to grab all the words
  componentDidMount() {
    fetch('/words')
      .then(res => res.json())
      .then(words => this.setState({ words }));
  }

  //this handle change updates the state every time the input field changes
  handleChange = e => {
    this.setState({
      inputVal: e.target.value
    });
  };

  //this is a simple swapping function that is used within the anagram finder
  swap(chars, a, b) {
    var temp = chars[a];
    chars[a] = chars[b];
    chars[b] = temp;
  }

  //in order to get the anagrams, a bit of logic is used in conjunction with the swap function above
  runAnagram() {
    let word = this.state.inputVal;
    var counter = [],
      anagrams = [],
      chars = word.split(''),
      length = chars.length,
      i;
    for (i = 0; i < length; i++) {
      counter[i] = 0;
    }
    anagrams.push(word);
    i = 0;
    while (i < length) {
      if (counter[i] < i) {
        this.swap(chars, i % 2 === 1 ? counter[i] : 0, i);
        counter[i]++;
        i = 0;
        anagrams.push(chars.join(''));
      } else {
        counter[i] = 0;
        i++;
      }
    }
 
    //in order to order by the second letter, i had to create two different constants, one filtering out the [1] letter and sorting it prior to the rest of the letters.
    //These anagrams are then returned sequentially down in the return statement
    const anagramsBySecondLetter = anagrams.filter(word => this.state.inputVal[1] === word[0]).sort().map((word, i) => <li className="list-group-item" key={i}>{word}</li>);
    const anagramsRest = anagrams.filter(word => this.state.inputVal[1] !== word[0]).sort().map((word, i) => <li className="list-group-item" key={i}>{word}</li>);
    const anagramLength = anagrams.length

    this.setState({
      display: `You have ${anagramLength} total anagrams`,
      anagramsBySecondLetter,
      anagramsRest
    });
  }

  //this handleClick function will look at the current word in the state (input field) and then run compare it to the other words in the dictionary. the anagram check will go here.
  handleClick = e => {
    let existsInArray = this.state.words.includes(this.state.inputVal);
    if (existsInArray === true) {
      this.runAnagram();
    } else {
      //the following sets the field of retrieved anagrams back to blanks
      //it only runs if a word is not found in the dictionary
      this.setState({
        display: 'None: Does Not Exist in Dictionary',
        anagramsBySecondLetter: '',
        anagramsRest: ''
      });
    }
  };

  render() {
    return (
      //React returns this simple JSX.
      <div style={{ paddingTop: '10%' }} className="App">
        <h1 className="card-title">Fluent Anagram Finder</h1>
        <input
          type="text"
          value={this.state.inputVal}
          onChange={this.handleChange}
        />
        <button className="btn btn-primary" onClick={this.handleClick}>Press for Anagrams</button>
        <h3>{this.state.inputVal}</h3>
        <p>{this.state.display}</p>
        {this.state.anagramsBySecondLetter}
        {this.state.anagramsRest}
      </div>
    );
  }
}

export default App;

