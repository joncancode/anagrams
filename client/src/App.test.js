import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('general <App/> component', () => {
  it('renders correctly without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('matches our snapshot', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('the methods', () => {
  it('should show an instance of componentDidMount', () => {
    const wrapper = shallow(<App />); 
    wrapper.instance().componentDidMount()
  });

  it('should show an instance of handleChange', () => {
    const wrapper = mount(<App />);
    const instance = wrapper.instance()
    instance.handleChange( { target: { value: "test input" } })
  });

  it('should show an instance of handleClick', () => {
    const wrapper = shallow(<App />); 
    wrapper.instance().handleClick()
  });

  it('should show an instance of runAlgorithm', () => {
    const wrapper = shallow(<App />); 
    wrapper.instance().runAnagram()
  });

});
