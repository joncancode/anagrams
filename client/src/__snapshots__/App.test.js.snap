// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`general <App/> component matches our snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="card-title"
        >
          Fluent Anagram Finder
        </h1>,
        <input
          onChange={[Function]}
          type="text"
          value=""
        />,
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          Press for Anagrams
        </button>,
        <h3>
          
        </h3>,
        <p />,
        undefined,
        undefined,
      ],
      "className": "App",
      "style": Object {
        "paddingTop": "10%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Fluent Anagram Finder",
          "className": "card-title",
        },
        "ref": null,
        "rendered": "Fluent Anagram Finder",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Press for Anagrams",
          "className": "btn btn-primary",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Press for Anagrams",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
      undefined,
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="card-title"
          >
            Fluent Anagram Finder
          </h1>,
          <input
            onChange={[Function]}
            type="text"
            value=""
          />,
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Press for Anagrams
          </button>,
          <h3>
            
          </h3>,
          <p />,
          undefined,
          undefined,
        ],
        "className": "App",
        "style": Object {
          "paddingTop": "10%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Fluent Anagram Finder",
            "className": "card-title",
          },
          "ref": null,
          "rendered": "Fluent Anagram Finder",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Press for Anagrams",
            "className": "btn btn-primary",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Press for Anagrams",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        undefined,
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
